transmat/git/git_unpack.go

```diff
@@ -0,0 +1,181 @@
+	if wareID.Type != PackType {
+		return api.WareID{}, Errorf(rio.ErrUsage, "this transmat implementation only supports packtype %q (not %q)", PackType, wareID.Type)
+	}
+	if placementMode == "" {
+		placementMode = rio.Placement_Copy
+	}
+	// Wrap the direct unpack func with cache behavior; call that.
+	return cache.Lrn2Cache(
+		osfs.New(config.GetCacheBasePath()),
+		unpack,
+	)(ctx, wareID, path, filt, placementMode, warehouses, mon)
+}
+
+func unpack(
+	ctx context.Context,
+	wareID api.WareID,
+	path string,
+	filt api.FilesetFilters,
+	placementMode rio.PlacementMode,
+	warehouses []api.WarehouseAddr,
+	mon rio.Monitor,
+) (_ api.WareID, err error) {
+	defer RequireErrorHasCategory(&err, rio.ErrorCategory(""))
+
+	// Sanitize arguments.
+	path2 := fs.MustAbsolutePath(path)
+	filt2, err := apiutil.ProcessFilters(filt, apiutil.FilterPurposeUnpack)
+	if err != nil {
+		return api.WareID{}, Errorf(rio.ErrUsage, "invalid filter specification: %s", err)
+	}
+
+	// Pick a warehouse and get a reader.
+	//  This is a *very* expensive operation for git.  It's less
+	//  of "pick a warehouse" and more "download the whole thing and hope we
+	//  get what we wanted" (which is very ironic for a system that has
+	//  a CAS system on its inside, yes).
+	whCtrl, err := pick(ctx,
+		wareID,
+		warehouses,
+		osfs.New(config.GetCacheBasePath().Join(fs.MustRelPath("git/objs"))),
+		mon,
+	)
+	if err != nil {
+		return api.WareID{}, err
+	}
+	tr, err := whCtrl.GetTree(wareID.Hash)
+	if err != nil {
+		panic(err)
+	}
+	tw := object.NewTreeWalker(tr, true, nil)
+
+	// Construct filesystem wrapper to use for all our ops.
+	afs := osfs.New(path2)
+
+	// Make the root dir.  Git doesn't have metadata for the tree root.
+	conjuredFmeta := fshash.DefaultDirMetadata()
+	filters.Apply(filt2, &conjuredFmeta)
+	if err := fsOp.PlaceFile(afs, conjuredFmeta, nil, filt2.SkipChown); err != nil {
+		return api.WareID{}, Errorf(rio.ErrInoperablePath, "error while unpacking: %s", err)
+	}
+
+	// Extract.
+	// Iterate over each entry, mutating filesystem as we go.
+	for {
+		fmeta := fs.Metadata{}
+		name, te, err := tw.Next()
+
+		// Check for done.
+		if err == io.EOF {
+			break // sucess!  end of archive.
+		}
+		if err != nil {
+			return api.WareID{}, Errorf(rio.ErrWareCorrupt, "corrupt git tree: %s", err)
+		}
+		if ctx.Err() != nil {
+			return api.WareID{}, Errorf(rio.ErrCancelled, "cancelled")
+		}
+		//fmt.Fprintf(os.Stderr, "walking git tree %s -- %#v\n", name, te)
+
+		// Reshuffle metainfo to our default format.
+		fmeta.Name = fs.MustRelPath(name)
+		switch te.Mode {
+		case filemode.Dir:
+			fmeta.Type = fs.Type_Dir
+			fmeta.Perms = 0644
+		case filemode.Regular:
+			fmeta.Type = fs.Type_File
+			fmeta.Perms = 0644
+		case filemode.Executable:
+			fmeta.Type = fs.Type_File
+			fmeta.Perms = 0755
+		case filemode.Symlink:
+			fmeta.Type = fs.Type_Symlink
+			fmeta.Perms = 0644
+			fmeta.Linkname = "TODO" // TODO ?!?!
+		case filemode.Submodule:
+			// TODO
```

Comment:
I'm still pretty adamant that we will need configurable recursion depth when we get to this.